### PR TITLE
Fixes #16483: nonblocking errors parser, more precise context capture

### PR DESCRIPTION
**Pull request by gpoblon**
Integration of a non-blocking error parser
Integration of another way to implement and catch error contexts allowing to catch more accurate informations about errors

https://issues.rudder.io/issues/16483

### CODE REVIEW

File: rudder-lang/src/parser.rs [hunk 1]
```diff
@@ -190,25 +191,47 @@ macro_rules! sequence {
...
+                };
                 let (i,_) = strip_spaces_and_comment(i)?;
             )*
             Ok((i, $output))
         }
     };
 }
 
+/// Tool function allowing to save last state of parsing offset line before any error
+fn get_focalized_input(i: PInput) -> PInput {
```
Comment (peckpeck) [inline rudder-lang/src/parser.rs]: Is focalized the right term, isn't there a more explicit word ?

File: rudder-lang/src/parser.rs [hunk 2]
```diff
@@ -190,25 +191,47 @@ macro_rules! sequence {
...
+    // Might be useful to expand via a second parameter the limit pattern or even a function to combine with
+    let err_result: nom::IResult<PInput, PInput> = take_until("\n")(i);
+    // return the received input by default
+    let (_next, err_holder) = err_result.unwrap_or((i, i));
```
Comment (peckpeck) [inline rudder-lang/src/parser.rs]: why name it err_holder ?

File: rudder-lang/src/parser.rs [hunk 3]
```diff
@@ -190,25 +191,47 @@ macro_rules! sequence {
...
             )*
             Ok((i, $output))
         }
     };
 }
 
+/// Tool function allowing to save last state of parsing offset line before any error
+fn get_focalized_input(i: PInput) -> PInput {
+    // Might be useful to expand via a second parameter the limit pattern or even a function to combine with
+    let err_result: nom::IResult<PInput, PInput> = take_until("\n")(i);
```
Comment (peckpeck) [inline rudder-lang/src/parser.rs]: why name it err_result ?

File: rudder-lang/src/parser.rs [hunk 4]
```diff
@@ -190,25 +191,47 @@ macro_rules! sequence {
...
         move |i| {
             $(
-                let (i,$f) = $parser (i)?;
+                // save last line (cursor is updated at parser calls). context of an potential error
+                let focalized_error = get_focalized_input(i);
+                // intercept error to update its context if it should lead to a handled compilation error
+                let (i, $f) = match $parser (i) {
+                    Ok(res) => res,
+                    Err(e) => return Err(update_error_context(e, focalized_error))
+                };
```
Comment (peckpeck) [inline rudder-lang/src/parser.rs]: you could also write it: 
```
  let (j,$f) ) ... {
       ... => return Err(update_error_context(e, get_focalized_input(i)))
  };
  let i = j;
```

Then you still have access to the original i and may some day find more information to extract from it

File: rudder-lang/src/parser.rs [hunk 5]
```diff
@@ -260,10 +283,10 @@ fn penum(i: PInput) -> PResult<PEnum> {
         {
             metadata: pmetadata_list; // metadata unsupported here, check done after 'enum' tag
             global: opt(tag("global")); // TODO at least one space here
-            e:      tag("enum"); // TODO at least one space here
-            _fail: or_fail(verify(peek(anychar), |_| metadata.is_empty()), || PErrorKind::UnsupportedMetadata(metadata[0].key.into()));
-            name:   or_fail(pidentifier, || PErrorKind::InvalidName(e));
-            b:      tag("{"); // do not fail here, it could still be a mapping
+            e:      or_err(tag("enum"), || PErrorKind::ExpectedReservedWord("enum")); // TODO at least one space here
+            _fail:  or_err(verify(peek(anychar), |_| metadata.is_empty()), || PErrorKind::UnsupportedMetadata(metadata[0].key.into()));
```
Comment (peckpeck) [inline rudder-lang/src/parser.rs]: why did you remove or_fail ? 
it seems to me that if the user has already written "enum blah" an invalid blah is a fail.
and if it is not, please rename the variable too

File: rudder-lang/src/parser.rs [hunk 6]
```diff
@@ -260,10 +283,10 @@ fn penum(i: PInput) -> PResult<PEnum> {
...
+            name:   or_err(pidentifier, || PErrorKind::InvalidName(e));
```
Comment (peckpeck) [inline rudder-lang/src/parser.rs]: This seems to be a fail too

File: rudder-lang/src/parser.rs [hunk 7]
```diff
@@ -967,6 +990,7 @@ pub enum PDeclaration<'src> {
     Alias(PAliasDef<'src>),
 }
 fn pdeclaration(i: PInput) -> PResult<PDeclaration> {
+    end_of_pfile(i)?;
```
Comment (peckpeck) [inline rudder-lang/src/parser.rs]: please add a comment on why this is here

File: rudder-lang/src/parser.rs [hunk 8]
```diff
@@ -977,6 +1001,17 @@ fn pdeclaration(i: PInput) -> PResult<PDeclaration> {
     ))(i)
 }
 
+fn end_of_pfile(i: PInput) -> PResult<()> {
```
Comment (peckpeck) [inline rudder-lang/src/parser.rs]: A little comment on what this parser does

File: rudder-lang/src/parser/error.rs [hunk 9]
```diff
@@ -166,30 +184,103 @@ where
...
-            Ok(y) => Ok(y),
+            Err(Err::Incomplete(_)) => panic!("Incomplete should never happen"),
+            Ok(y) => Ok(y)
         }
     }
 }
 
+/// This function turns our own `Error`s (not nom ones) into `Failure`s so they are properly handled
+/// by the `nom::multi::many0()` function which abstracts Errors, only breaking on failures which was an issue
+pub fn or_fail_perr<'src, O, F>(f: F) -> impl Fn(PInput<'src>) -> PResult<'src, O>
```
Comment (peckpeck) [inline rudder-lang/src/parser/error.rs]: i think there is a better name, "_perr" here is emaningless

File: rudder-lang/src/parser/error.rs [hunk 10]
```diff
@@ -166,30 +184,103 @@ where
...
     move |input| {
-        let x = f(input);
-        match x {
+        match f(input) {
             // a non nom error cannot be superseded
             // keep original context when possible
             Err(Err::Failure(err)) => match err.kind {
                 PErrorKind::Nom(_) => Err(Err::Failure(PError {
-                    context: err.context,
+                    context: None,
```
Comment (peckpeck) [inline rudder-lang/src/parser/error.rs]: why are you removing the original context ? 
it may contain useful data, especially if you forget or don't need to override it

File: rudder-lang/src/parser/error.rs [hunk 11]
```diff
@@ -166,30 +184,103 @@ where
...
             })),
-            Err(Err::Incomplete(_)) => Err(Err::Failure(PError {
-                context: input,
+            Err(Err::Incomplete(_)) => panic!("Incomplete should never happen"),
+            Ok(y) => Ok(y),
+        }
+    }
+}
+
+/// Similar code than `or_fail()` yet usage and behavior differ:
```
Comment (peckpeck) [inline rudder-lang/src/parser/error.rs]: similar .. to

File: rudder-lang/src/parser/error.rs [hunk 12]
```diff
@@ -166,30 +184,103 @@ where
...
+/// primarly it is non-blockant as it does not turn Errors into Failures
```
Comment (peckpeck) [inline rudder-lang/src/parser/error.rs]: non blocking

Comment (peckpeck) [reply rudder-lang/src/parser/error.rs]: the proper should be non terminating

File: rudder-lang/src/parser/error.rs [hunk 13]
```diff
@@ -166,30 +184,103 @@ where
...
+where
+    F: Fn(PInput<'src>) -> PResult<'src, O>,
+    E: Fn() -> PErrorKind<PInput<'src>>,
+{
+    move |input| {
+        match f(input) {
+            Err(Err::Failure(err)) => match err.kind {
+                PErrorKind::Nom(_) => Err(Err::Failure(PError {
+                    context: None,
+                    kind: e(),
```
Comment (peckpeck) [inline rudder-lang/src/parser/error.rs]: i'm not sure we should override the error kind here, could you give an example ?

File: rudder-lang/src/parser/error.rs [hunk 14]
```diff
@@ -166,30 +184,103 @@ where
...
+                })),
+                _ => Err(Err::Failure(e)),
+            },
+            Err(Err::Incomplete(_)) => panic!("Incomplete should never happen"),
+            Ok(y) => Ok(y)
+        }
+    }
+}
+
+/// Updates content of on error to fit and capture a better context
```
Comment (peckpeck) [inline rudder-lang/src/parser/error.rs]: "on error" ?

File: rudder-lang/src/parser/error.rs [hunk 15]
```diff
@@ -166,30 +184,103 @@ where
...
+/// Solely exists for (w)sequence macro
+pub fn update_error_context<'src>(e: Err<PError<PInput<'src>>>, new_ctx: PInput<'src>) -> Err<PError<PInput<'src>>> {
+    match e {
+        Err::Failure(err) => {
+            let context = match err.context {
+                None => Some(new_ctx),
+                Some(context_to_keep) => Some(context_to_keep)
+            };
+            Err::Failure(PError {
+                context: context,
```
Comment (peckpeck) [inline rudder-lang/src/parser/error.rs]: this can be written " context, "

File: rudder-lang/src/parser/error.rs [hunk 16]
```diff
@@ -123,27 +124,44 @@ impl<I: Clone> ParseError<I> for PError<I> {
...
-            self.context.fragment,
-            "-->".bright_blue(),
-            message.bold(),
-            
-        ))
+
+        // simply removes superfluous line return (prettyfication)
+        match self.context {
+            Some(ctx) => {
+                let mut context = ctx.fragment.to_owned();
```
Comment (amousset) [inline rudder-lang/src/parser/error.rs]: Using https://doc.rust-lang.org/std/primitive.str.html#method.trim_end_matches would avoid creating a new mutable string, but it's not mandatory and current solution is acceptable.

File: rudder-lang/src/parser.rs (hunk 5 above)
Comment (gpoblon) [reply rudder-lang/src/parser.rs]: Just checked, this was a mistake: I needed to have every enum error non-terminated for testing purpose and forgot to revert it. Fixed

File: rudder-lang/src/parser.rs (hunk 3 above)
Comment (gpoblon) [reply rudder-lang/src/parser.rs]: I wanted to make it explicit that it is only call it whenever we hit an error.
But yeah, I've updated the entire function naming so it's not ambiguous

File: rudder-lang/src/parser.rs (hunk 1 above)
Comment (gpoblon) [reply rudder-lang/src/parser.rs]: renamed

File: rudder-lang/src/parser/error.rs (hunk 16 above)
Comment (gpoblon) [reply rudder-lang/src/parser/error.rs]: nice tip, thanks, updated

File: rudder-lang/src/parser/error.rs (hunk 13 above)
Comment (gpoblon) [reply rudder-lang/src/parser/error.rs]: Do you mean that it is just superfuous code since anyway `or_fail()` will be called in case of a failure ?

Comment (gpoblon) [reply rudder-lang/src/parser/error.rs]: fixed, it was superfluous indeed

**Comment by gpoblon** [comment]
All fixes related to previous comments pushed

File: rudder-lang/src/parser.rs [hunk 17]
```diff
@@ -209,27 +208,26 @@ macro_rules! wsequence {
...
-                let focalized_error = get_focalized_input(i);
                 // intercept error to update its context if it should lead to a handled compilation error
-                let (i, $f) = match $parser (i) {
+                let (j, $f) = match $parser (i) {
                     Ok(res) => res,
-                    Err(e) => return Err(update_error_context(e, focalized_error))
+                    Err(e) => return Err(update_error_context(e, get_accurate_context(i)))
                 };
-                let (i,_) = strip_spaces_and_comment(i)?;
+                let (j,_) = strip_spaces_and_comment(j)?;
```
Comment (peckpeck) [inline rudder-lang/src/parser.rs]: let (i,_)

Comment (gpoblon) [reply rudder-lang/src/parser.rs]: Corrigé
